src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -197,6 +197,23 @@ private void TagBlock(Stack<Tuple<HtmlSymbol, SourceLocation>> tags)
                         // End Tag
                         return EndTag(tagStart, tags, tagBlockWrapper);
                     case HtmlSymbolType.Bang:
+                        var nextSymbol = Lookahead();
+
+                        // If the next symbol is a text symbol and is not DOCTYPE then it's an ordinary HTML tag.
+                        if (nextSymbol.Type == HtmlSymbolType.Text &&
+                            !string.Equals(nextSymbol.Content, "DOCTYPE", StringComparison.OrdinalIgnoreCase))
+                        {
+                            using (tagBlockWrapper = Context.StartBlock(BlockType.Tag))
+                            {
+                                var complete = StartTag(tags, tagBlockWrapper);
+
+                                // If the tag was completed this will ensure the '>' is contained in the tag block.
+                                Output(SpanKind.Markup);
+
+                                return complete;
+                            }
```

Comment:
I had that initially but quickly found out that for this code path (expectation = at a special tag = `!`) the `tagBlockWrapper` that's passed in is `null` (ref https://github.com/aspnet/Razor/blob/disableths.187/src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs#L136) and therefore nothing gets encapsulated in a `TagBlock` unless we manually encapsulate it.

I really dislike the parser, I was face palming every character of this snippet :sob: